### PR TITLE
[Enhancement] Support backup restore for logical view

### PR DESCRIPTION
**Pull request by srlch**
Support backup restore for logical view

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/backup/RestoreJob.java [hunk 1]
```diff
@@ -527,6 +544,15 @@ private void checkAndPrepareMeta() {
...
                             .getTable(db.getFullName(), jobInfo.getAliasByOriginNameIfSet(tblInfo.name));
                 if (localTbl != null) {
+                    if (remoteTbl.isOlapView() && !localTbl.isOlapView()) {
+                        status = new Status(ErrCode.BAD_REPLACE,
+                                            "Table: " + tblInfo.name + " has existed and it is not a View");
+                        return;
+                    } else if (remoteTbl.isOlapView()) {
+                        restoredOlapViews.put((View) remoteTbl, true);
+                        continue;
+                    }
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/backup/RestoreJob.java:554]: there are 2 `remoteTbl.isOlapView()`, could refactor the code.

File: fe/fe-core/src/main/java/com/starrocks/backup/RestoreJob.java [hunk 2]
```diff
@@ -832,6 +869,60 @@ protected void sendCreateReplicaTasks() {
...
+                } catch (Exception e) {
+                    status = new Status(ErrCode.COMMON_ERROR, "failed to init View in restore" +
+                                        restoredOlapView.getName() + " " + e.getMessage());
+                    return;
+                }
+
+                try {
+                    globalStateMgr.getLocalMetastore().onCreate(db, view, "", false);
+                } catch (DdlException e) {
+                    status = new Status(ErrCode.COMMON_ERROR, "failed to create View in restore" +
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/backup/RestoreJob.java:917]: better to use `LocalMetastore.createView()` to create or alter view.
need refactor `LocalMetastore.createView()` first to reuse the code.

File: fe/fe-core/src/main/java/com/starrocks/backup/BackupJob.java [hunk 3]
```diff
@@ -473,14 +477,17 @@ private void prepareAndSendSnapshotTask() {
             // create snapshot tasks
             for (TableRef tblRef : tableRefs) {
                 String tblName = tblRef.getName().getTbl();
-                OlapTable tbl = (OlapTable) globalStateMgr.getLocalMetastore()
-                            .getTable(db.getFullName(), tblName);
+                Table tbl = globalStateMgr.getLocalMetastore().getTable(db.getFullName(), tblName);
+                if (tbl.isOlapView()) {
+                    continue;
+                }
+                OlapTable olapTbl = (OlapTable) tbl;
```
[the remark this answers is not in the document]
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/backup/BackupJob.java:484]: We need to get the partition info to backup

File: fe/fe-core/src/main/java/com/starrocks/backup/RestoreJob.java (hunk 1 above)
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/backup/RestoreJob.java]: fixed

File: fe/fe-core/src/main/java/com/starrocks/backup/BackupJobInfo.java [hunk 4]
```diff
@@ -147,6 +147,8 @@ public static class BackupTableInfo {
         public Long autoIncrementId;
         @SerializedName(value = "partitions")
         public Map<String, BackupPartitionInfo> partitions = Maps.newHashMap();
+        @SerializedName(value = "tableType")
+        public int tableType;
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/backup/BackupJobInfo.java:151]: Why not use Table.TableType ?

File: fe/fe-core/src/main/java/com/starrocks/backup/BackupJobInfo.java [hunk 5]
```diff
@@ -449,8 +457,19 @@ private static void genFromJson(String json, BackupJobInfo jobInfo) {
...
                 tblInfo.autoIncrementId = null;
             }
+            try {
+                tblInfo.tableType = tbl.getInt("tableType");
+            } catch (Exception e) {
+                tblInfo.tableType = Table.TableType.OLAP.ordinal();
+            }
             JSONObject parts = tbl.getJSONObject("partitions");
             String[] partsNames = JSONObject.getNames(parts);
+            Table.TableType tblType = Table.TableType.values()[tblInfo.tableType];
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/backup/BackupJobInfo.java:467]: What if no tableType in the table info in old version ?

File: fe/fe-core/src/main/java/com/starrocks/backup/RestoreJob.java [hunk 6]
```diff
@@ -832,6 +869,30 @@ protected void sendCreateReplicaTasks() {
         }
     }
 
+    protected void addRestoreOlapView(List<View> restoredOlapViews) {
+        Database db = globalStateMgr.getLocalMetastore().getDb(dbId);
+
+        ConnectContext context = new ConnectContext();
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/backup/RestoreJob.java:875]: the context is not used

File: fe/fe-core/src/main/java/com/starrocks/backup/BackupJobInfo.java (hunk 5 above)
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/backup/BackupJobInfo.java]: Updated, if restore older version of snapshot, treat the type as Table.TableType.OLAP is fine.

File: fe/fe-core/src/main/java/com/starrocks/backup/BackupJobInfo.java (hunk 4 above)
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/backup/BackupJobInfo.java:151]: `org.json.JSONObject` seems not support serialization and deserialization for enum type. It should be convert to INT or String to do the serialization and deserialization

File: fe/fe-core/src/main/java/com/starrocks/backup/RestoreJob.java (hunk 6 above)
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/backup/RestoreJob.java:875]: `setThreadLocalInfo` will set the connectcontext as thread local varaible and will be used in `createView`

Comment (xiangguangyxg) [reply fe/fe-core/src/main/java/com/starrocks/backup/RestoreJob.java:875]: Is those fileds enough for `createView()` ?

Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/backup/RestoreJob.java:875]: Fixed, add sqlmode in connectContext

File: fe/fe-core/src/main/java/com/starrocks/backup/BackupJob.java [hunk 7]
```diff
@@ -509,12 +516,16 @@ private void prepareAndSendSnapshotTask() {
             List<Table> copiedTables = Lists.newArrayList();
             for (TableRef tableRef : tableRefs) {
                 String tblName = tableRef.getName().getTbl();
-                OlapTable tbl = (OlapTable) globalStateMgr.getLocalMetastore()
-                            .getTable(db.getFullName(), tblName);
+                Table tbl = globalStateMgr.getLocalMetastore().getTable(db.getFullName(), tblName);
+                if (tbl.isOlapView()) {
+                    copiedTables.add(tbl);
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/backup/BackupJob.java:521]: not copied，is there consistent problem ？

Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/backup/BackupJob.java]: Updated, deep copy is prefer

**Review by xiangguangyxg — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Pull request merged**